proposals/0000-componentized-cabal.rst

```diff
@@ -0,0 +1,381 @@
+
+2. It makes it possible for package managers like cabal-install and
+   Stack to organize builds on a per-component basis, improving
+   parallelism (e.g., when a package contains many independent
+   executables) and making it easier to avoid unnecessary rebuilds when
+   a new component is built. (`#2775 <https://github.com/haskell/cabal/issues/2775>`_)
+
+3. It solves the cyclic dependency problem where a test suite depends on
+   a testing library that depends on the the library being tested. Now
+   the test suite can be built independently from the library, after the
+   testing library has been built.
+
+4. It allows for an architecturally sound implementation of Backpack, as
+   the process of instantiating requirements of a Backpack only makes
+   sense for libraries, rather than an entire package (an executable and
+   test suite cannot be instantiated in any meaningful sense).
+
+I have divided this proposal into two parts: first the changes necessary
+for the Cabal library (which is shared between cabal-install and Stack),
+and then a description of how cabal-install can take advantage of this
+new feature.  There is not a Stack portion to this proposal, but
+commentary from the Stack team would be appreciated.
+
+Proposed Change: Cabal
+----------------------
+
+**The status quo.** Currently, the Cabal library defines an command line
+API via the ``./Setup configure`` command. A conventional invocation of
+this command from, e.g., ``cabal-install``, might look like this::
+
+    ./Setup configure --ghc --prefix=$HOME/.cabal --user --ipid=package-name-1.0-abcdefg --dependency=array=array-0.5.1.0-abcdef1234567 --dependency=base=base-4.8.2.0-abcdef123456 --dependency=tasty=tasty-1.0-abcdef123456 --enable-tests --exact-configuration --disable-benchmarks
+
+There are a few important flags: ``--ipid`` specifies the *installed
+package identity* that the package will be built under: this is
+identifier is a unique identifier chosen by the package manager (for
+example, ``cabal new-build`` computes this identifier by hashing all of
+the possible inputs into the build, with the hope that the hash uniquely
+determines all of the build products of the package), and
+``--dependency=pkgname=ipid-1.0-abcdef`` specifies a specific dependency
+to be used during the build of all the components.
+``--exact-configuration`` requires ALL dependencies of all enabled
+components to be specified explicitly.  Notice that this particular
+command has ``--enable-tests`` passed (thus enabling the building of the
+test suite) and specifies which version of ``tasty`` (a test library)
+should be used. This command configures both the library and the test
+suite of this package *simultaneously*.
+
+Currently, the ``configure`` command takes no positional arguments.
+
+**The proposal.** We extend ``configure`` to take a single argument
+specifying which component is to be configured::
+
+    ./Setup configure test-suite
```

Comment:
Is there any rationale for this approach rather than using a flag to indicate this new mode, like we do for `--exact-configuration`.